### ISSUE TITLE
Got divided by zero PHP Warning while getting exponential random numbers.

### ISSUE THREAD

**Issue by greensea** [markrogoyski/math-php#269]
I am using v0.43.0.
I try to get exponential random numbers, but sometimes I got divided by zero warnings, and the rand() function give me a NAN.
To reproduce this issue, try the code:

```
use MathPHP\Probability\Distribution\Continuous;

for ($i =0; $i < 100; $i++) {
    $λ           = 1.0; // rate parameter
    $exponential = new Continuous\Exponential($λ);
    $r           = $exponential->rand();
    printf("%s\n", $r);
}
```

Then you will get divided by zero warnings.

> Warning:  Division by zero in ****/vendor/markrogoyski/math-php/src/Probability/Distribution/Continuous/Continuous.php on line 37

**Comment by markrogoyski** [comment]
Hi @greensea,
Thank you for your interest in MathPHP and reporting an issue. I was able to reproduce it with the code you provided. Thank you.

I suspect the issue is in the Continuous inverse where the PDF is being computed and a result of 0 is being used. We will investigate and report back when there is a solution.

Thanks again for reporting the issue.
Mark

**Comment by peter279k** [comment]
I have the same issue about this when using the `0.43.0` version. And It seems that the following code has this problem about the line `41` from this message:` php/src/Probability/Distribution/Continuous/Continuous.php on line 41`.

```php
$guess = $del_y / $slope + $guess;
```
I think the `$slope` variable will be zero in some conditions.

How about using the `if` condition to check whether the `$slope` variable is zero?

**Comment by markrogoyski** [comment]
Hi @peter279k,
For a short term solution I may implement something like that. Longer term I may investigate another solution.

Sorry for the delay in getting this fixed. I will prioritize this and get it done soon.

**Comment by peter279k** [comment]
@markrogoyski, thank you for your reply. After tracing the source code deeply, I found where the problem is.

In the line `39` of `/src/Probability/Distribution/Continuous/Continuous.php`, the following code is:

```
$slope = $this->pdf($guess);
```

After checking the `pdf` method in `Exponential` class, the `$guess` will be the zero because the `$guess` will be the negative value in some cases. Here is the `pdf` method:

```php
    public function pdf(float $x): float
    {
        if ($x < 0) {
            return 0;
        }
        Support::checkLimits(self::SUPPORT_LIMITS, ['x' => $x]);

        $λ = $this->λ;

        return $λ * exp(-$λ * $x);
    }
```

After checking the `$guess` variable, it will get the negative value in some test cases.

I set the break point in `Continuous` class to debug this issue when the `$guess` variable is negative value.

```php
while ($dif > $tolerance) {
    if ($guess <= 0) { var_dump($guess); var_dump($del_y); var_dump($target);  die();}
......
```

```
/home/lee/math-php/src/Probability/Distribution/Continuous/Continuous.php:36:
double(-0.19831428661946)
/home/lee/math-php/src/Probability/Distribution/Continuous/Continuous.php:36:
double(-0.44083519010932)
/home/lee/math-php/src/Probability/Distribution/Continuous/Continuous.php:36:
double(0.19128536871924)
```

I think it's the `pdf` issue and it should consider the `$guess` variable :+1: .

Thanks.

**Comment by markrogoyski** [comment]
@Beakerboy,
The way we are computing the inverse of a distribution will result in a divide by zero error if the PDF ever returns 0. The inverse is working fine for a lot of distributions and there are unit tests on those, so it seems like it is working OK most of the time, but for the Exponential distribution it seems to be having this issue because of the 0 PDF result.

Is the method we are using able to compute inverses for all distributions? Or do we need special-case handling for certain distributions? If you have a fix, or can suggest a solution, let me know. Otherwise, I will have to research this a bit before offering up a good fix.

@peter279k, @greensea,
Sorry for the delay in resolving this issue. I want to make sure it is fixed properly and has good regression tests once a fix is in place. Thank you for your patience.

**Comment by Beakerboy** [comment]
The exponential has a closed form inverse which should be added to the class:
X=ln(1-p)/-lambda

This will resolve the problem for this particular distribution.
What I don’t understand is why it would return a zero slope in the first place. If you look at the shape of the CDF, there is no min or max.

**Comment by markrogoyski** [comment]
For the meantime I'll add the closed-form inverse to the Exponential class. Thanks for looking into it.

**Comment by markrogoyski** [comment]
This has been fixed in the latest release, [v0.44.0](https://github.com/markrogoyski/math-php/releases/tag/v0.44.0).

The exponential distribution now uses the closed-form inverse function.

Thank you for your patience in getting this fix out. And thank you for your continued interest in MathPHP!

@Beakerboy, thanks for the suggestion of using the close-form function.
